src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/TempDataDictionary.cs

```diff
@@ -11,12 +11,14 @@ namespace Microsoft.AspNetCore.Mvc.ViewFeatures
     /// <inheritdoc />
     public class TempDataDictionary : ITempDataDictionary
     {
+        // Perf: Everything here is lazy because the TempDataDictionary is frequently created and passed around
+        // without being manipulated.
         private Dictionary<string, object> _data;
         private bool _loaded;
         private readonly ITempDataProvider _provider;
         private readonly HttpContext _context;
-        private HashSet<string> _initialKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-        private HashSet<string> _retainedKeys = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private HashSet<string> _initialKeys;
+        private HashSet<string> _retainedKeys;
```

Comment:
This is safe because every actual method either calls `Load()` or checks `_isLoaded` and no-ops.